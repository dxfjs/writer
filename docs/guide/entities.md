---
layout: doc
---

# Entities

Entities are the graphical objects in dxf.

:::info

All entities described here are added directly to the model space block.

:::

All Entities have some common options:

- `trueColor`
- `colorNumber`
- `layerName`
- `visible`
- `lineType`
- `lineTypeScale`

## `ARC` entity

An arc is a portion of the circumference of a circle.
(See more in [Wikipedia](https://en.wikipedia.org/wiki/Circular_arc))

```js
import DxfWriter, { point3d, addArc } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addArc(point3d(0, 0, 0), 10, 0, 45);
// or
const myArc = addArc(point3d(0, 0, 0), 10, 0, 45); // If you want a reference to the added arc.
```

## `CIRCLE` entity

A circle is a shape consisting of all points in a plane that are at a given distance from a given point, the centre.
Equivalently, it is the curve traced out by a point that moves in a plane so that its distance from a given point is constant.
The distance between any point of the circle and the centre is called the radius. (From [Wikipedia](https://en.wikipedia.org/wiki/Circle))

```js
import DxfWriter, { point3d, addCircle } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addCircle(point3d(0, 0, 0), 10);
// or
const myCircle = addCircle(point3d(0, 0, 0), 10); // If you want a reference to the added circle.
```

## `ELLIPSE` entity

An ellipse is a plane curve surrounding two focal points, such that for all points on the curve, the sum of the two distances to the focal points is a constant.
As such, it generalizes a circle, which is the special type of ellipse in which the two focal points are the same.
(From [Wikipedia](https://en.wikipedia.org/wiki/Ellipse))

```js
import DxfWriter, { point3d, addEllipse } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addEllipse(point3d(100, 100, 0), point3d(50, 0, 0), 0.5, 0, 2 * Math.PI);

// or if you want a reference to the added ellipse.
const myEllipse = addEllipse(
  point3d(100, 100, 0),
  point3d(50, 0, 0),
  0.5,
  0,
  2 * Math.PI
);
```

![ellipse demo](_media/ellipse-demo.png)

## `3DFACE` entity

A 3d face is flat surface with 4 corners.

```js
import DxfWriter, { point3d, add3dFace } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.add3dFace(
  point3d(0, 0, 0),
  point3d(0, 100, 0),
  point3d(100, 100, 0),
  point3d(100, 0, 0)
);

// or if you want a reference to the added face.
const my3dFace = add3dFace(
  point3d(0, 0, 0),
  point3d(0, 100, 0),
  point3d(100, 100, 0),
  point3d(100, 0, 0)
);
```

You can contole the visibilty of the edges of the 3d Face by using the convenience methods:

```js
my3dFace.setFirstEdgeVisible(true);
my3dFace.setSecondEdgeVisible(true);
my3dFace.setThirdEdgeVisible(true);
my3dFace.setFourthEdgeVisible(true);

//Or all of them at once
my3dFace.setEdgesVisible(true);
```

> By default all of them are visible.

Also you can set the visibilty in the options:

```js
import DxfWriter, {
  point3d,
  add3dFace,
  InvisibleEdgeFlags,
} from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.add3dFace(
  point3d(0, 0, 0),
  point3d(0, 100, 0),
  point3d(100, 100, 0),
  point3d(100, 0, 0),
  {
    invisibleEdges: InvisibleEdgeFlags.First | InvisibleEdgeFlags.Fourth,
  }
);
// This will make the first and fourth edges invisible.
```

:::tip

To define a 3d face with only 3 corners, make the fourth same as the third corner ðŸ˜‰.

:::

## `HATCH` entity

The HATCH entity fills an enclosed area defined by one or more boundary paths with a hatch pattern : solid or gradient.

To create a hatch you need to define :

- A boundary path or multiple.
  - Polyline path is the simple way
  - Edges path is the flexible way
- A hatch pattern solid or gradient.

Define a polyline path :

```js
import { HatchPolylineBoundary, vertex } from "@tarikjabiri/dxf";

const polyline = new HatchPolylineBoundary();
polyline.add(vertex(0, 0));
polyline.add(vertex(0, 10000));
polyline.add(vertex(10000, 10000));
polyline.add(vertex(10000, 0));
// You can define a bulge for the vertex also
polyline.add(vertex(10000, 0, 10));
```

Define a edges path :

```js
import { HatchEdgesTypeData, point2d } from "@tarikjabiri/dxf";

const edges = new HatchEdgesTypeData();
edges.addLineEdgeData(point2d(0, 0), point2d(0, 10000));
edges.addLineEdgeData(point2d(0, 10000), point2d(10000, 10000));
edges.addLineEdgeData(point2d(10000, 10000), point2d(10000, 0));
edges.addLineEdgeData(point2d(10000, 0), point2d(0, 0));
// For now LineEdge and ArcEdge are supported
```

Create a `HatchBoundaryPath` instance :

```js
import { HatchBoundaryPath } from "@tarikjabiri/dxf";

const boundary = new HatchBoundaryPath();

// Add the defined path
boundary.addPolylineBoundary(polyline); // You can add multiple
// Or
boundary.addEdgesTypeData(edges); // You can add multiple
```

:::warning
You can define one type of paths `HatchPolylineBoundary` or `HatchEdgesTypeData`, not both at same time.
:::

Next you define the fill pattern solid or gradient :

```js
import { pattern, gradient, GradientType } from "@tarikjabiri/dxf";
// Solid
const mysolid = pattern({
  name: HatchPredefinedPatterns.STEEL,
  // Other properties you can define optionally
  // angle?: number;
  // scale?: number;
  // double?: boolean;
});

// Gradient
const mygradient = gradient({
  firstColor: 5,
  type: GradientType.CYLINDER,
  // Other properties you can define optionally
  // secondColor?: number;
  // angle?: number;
  // definition?: number;
  // tint?: number;
});
```

Now you have a boundary path and a fill pattern, you can add the hatch :

```js
import DxfWriter, { addHatch } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

// You can call addHatch from dxf object
dxf.addHatch(boundary, mygradient);

// Or call directly addHatch to store a reference to the hatch object
const hatch = addHatch(boundary, mygradient);
```

:::info
This way you can add only non-associative hatches.
:::

## `IMAGE` entity

```js
import DxfWriter, { point3d, addImage } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addImage(
  ".\\image-name.png", // Or the absolute path of the image if it isn't int the same folder.
  "image-name",
  point3d(462419.04, 576568.45, 0), // Insertion point of the bottomLeft corner of the image.
  1792, // the width of the image
  1280, // the height of the image
  1, // Scale
  0 // rotation
);
```

If you want a reference to the added image use `addImage()`.

## `INSERT` entity

It's convenient to create reusable blocks at once and use them all over the places you want.
After defining your blocks, you can insert them with the `INSERT` entity. See example code below:

```js
import DxfWriter, { point3d, addInsert } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

// Define your block
const myBlock = addBlock("block1");
myBlock.addCircle(point3d(0, 0, 0), 50);
myBlock.addRectangle(point2d(-35.3553, 35.3553), point2d(35.3553, -35.3553));

// Insert it when and where ever you want.
dxf.addInsert(myBlock.name, point3d(0, 0, 0));
// Or if you want a reference to the added insert.
const myInsert = addInsert(myBlock.name, point3d(0, 0, 0));
```

## `LINE` entity

```js
import DxfWriter, { point3d, addLine } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addLine(point3d(0, 0, 0), point3d(100, 100, 0));
// Or if you want a reference to the added line.
const line = addLine(point3d(0, 0, 0), point3d(100, 100, 0));
```

## `LWPOLYLINE` entity

```js
import DxfWriter, { point3d, addLine } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

const vertices = [
  {
    point: point2d(0, 0, 0),
  },
  {
    point: point2d(100, 100, 0),
  },
  {
    point: point2d(100, 200, 0),
  },
  {
    point: point2d(0, 300, 0),
  },
];

dxf.addLWPolyline(vertices, {
  flags: LWPolylineFlags.Closed,
});

// Or if you want a reference to the added lwpolyline.
const lwpolyline = addLWPolyline(vertices);
```

## `POINT` entity

```js
import DxfWriter, { point3d, addPoint } from "@tarikjabiri/dxf";

const dxf = new DxfWriter();

dxf.addLine(point3d(0, 0, 0), point3d(100, 100, 0));
// Or if you want a reference to the added line.
const line = addLine(point3d(0, 0, 0), point3d(100, 100, 0));
```

## `POLYLINE` entity

## `SPLINE` entity

## `TEXT` entity
