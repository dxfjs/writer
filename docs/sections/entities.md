# Entities

All entities described here can be added to the model space, paper spaces or blocks.

```ts
const modelSpace = writer.document.modelSpace;
const paperSpace = writer.document.paperSpace;
const myBlock = writer.document.blocks.addBlock({
  name: "myBlock",
});
```

## Line

```ts
const line = modelSpace.addLine({
  start: point(),
  end: point(100, 100),
});
```

## Point

```ts
const point = modelSpace.addPoint({
  x: 5,
  y: 5,
});
```

## Ray

```ts
const line = modelSpace.addRay({
  start: point(),
  unitDirectionVector: point(10, 10),
});
```

## Spline

```ts
const spline = modelSpace.addSpline({
  controls: [
    point(),
    point(10, 10),
    point(20, 10),
    point(30, 20),
    point(100, 100),
  ],
});
```

## Text

```ts
const text = modelSpace.addText({
  value: "Hello World!",
  firstAlignmentPoint: point(),
  height: 10,
});
```
