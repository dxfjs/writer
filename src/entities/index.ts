export * from "./arc";
export * from "./circle";
export * from "./dimension";
export * from "./ellipse";
export * from "./entities";
export * from "./entity";
export * from "./face";
export * from "./hatch";
export * from "./insert";
export * from "./leader";
export * from "./line";
export * from "./lwpolyline";
export * from "./manager";
export * from "./mesh";
export * from "./mleader";
export * from "./mtext";
export * from "./point";
export * from "./polyline";
export * from "./ray";
export * from "./seqend";
export * from "./solid";
export * from "./spline";
export * from "./table";
export * from "./text";
export * from "./vertex";
