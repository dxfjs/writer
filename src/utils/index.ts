export * from "./application";
export * from "./bbox";
export * from "./color";
export * from "./constants";
export * from "./functions";
export * from "./handle";
export * from "./tags";
export * from "./xdata";
