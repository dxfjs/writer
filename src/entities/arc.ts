import { BBox, BoundingBox, Handle, TagsManager } from "../utils";
import { Entity, EntityOptions } from "./entity";
import { Point3D } from "../types";

export interface ArcOptions extends EntityOptions {
  thickness?: number;
  center: Point3D;
  radius: number;
  startAngle: number;
  endAngle: number;
}

export class Arc extends Entity {
  thickness?: number;
  center: Point3D;
  radius: number;
  startAngle: number;
  endAngle: number;

  override get subClassMarker(): string {
    return "AcDbCircle";
  }

  constructor(options: ArcOptions, handle: Handle) {
    super("ARC", handle, options);
    this.thickness = options.thickness;
    this.center = options.center;
    this.radius = options.radius;
    this.startAngle = options.startAngle;
    this.endAngle = options.endAngle;
  }

  override bbox(): BoundingBox {
    return BBox.point(this.center);
  }

  protected override tagifyChild(mg: TagsManager): void {
    mg.add(39, this.thickness);
    mg.point(this.center);
    mg.add(40, this.radius);
    mg.add(100, "AcDbArc");
    mg.add(50, this.startAngle);
    mg.add(51, this.endAngle);
  }
}
