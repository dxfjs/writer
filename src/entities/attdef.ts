import { Handle, TagsManager } from "@/utils";
import { Text, TextOptions } from "./text";

export interface AttdefOptions extends TextOptions {
  prompt?: string;
  tag: string;
  flags?: number;
}

export class Attdef extends Text {
  prompt: string;
  tag: string;
  flags: number;

  protected override get subClassMarker2(): string {
    return "AcDbAttributeDefinition";
  }

  constructor(options: AttdefOptions, handle: Handle) {
    super(options, handle);
    this._type = "ATTDEF";
    this.prompt = options.prompt || "";
    this.tag = options.tag;
    this.flags = options.flags || 0;
  }

  protected override tagifyChild(mg: TagsManager): void {
    mg.add(39, this.thickness);
    mg.point(this.firstAlignmentPoint);
    mg.add(40, this.height);
    mg.add(1, this.value);
    mg.add(50, this.rotation);
    mg.add(41, this.relativeXScaleFactor);
    mg.add(51, this.obliqueAngle);
    mg.add(7, this.styleName);
    mg.add(71, this.generationFlags);
    mg.add(72, this.horizontalJustification);
    mg.point(this.secondAlignmentPoint, 1);
    mg.add(100, this.subClassMarker2);
    mg.add(280, 0);
    mg.add(3, this.prompt);
    mg.add(2, this.tag);
    mg.add(70, this.flags);
    mg.add(74, this.verticalJustification);
    mg.add(280, 1);
  }
}
