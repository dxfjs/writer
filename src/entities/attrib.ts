import { Entity, EntityOptions } from "./entity";
import { Handle, TagsManager, extrusion } from "@/utils";
import { Point3D, Union } from "@/types";
import { TextHorizontalJustification, TextVerticalJustification } from "./text";
import { Insert } from "./insert";
import { TextGenerationFlags } from "@/tables";

export interface AttribOptions extends EntityOptions {
  thickness?: number;
  startPoint: Point3D;
  height: number;
  value: string;
  tag: string;
  flags?: number;
  rotation?: number;
  relativeXScaleFactor?: number;
  obliqueAngle?: number;
  styleName?: string;
  generationFlags?: Union<typeof TextGenerationFlags>;
  horizontalJustification?: Union<typeof TextHorizontalJustification>;
  verticalJustification?: Union<typeof TextVerticalJustification>;
  alignmentPoint?: Point3D;
  extrusion?: Point3D;
  insert: Insert;
}

export class Attrib extends Entity {
  thickness?: number;
  startPoint: Point3D;
  height: number;
  value: string;
  tag: string;
  flags: number;
  rotation?: number;
  relativeXScaleFactor?: number;
  obliqueAngle?: number;
  styleName?: string;
  generationFlags?: Union<typeof TextGenerationFlags>;
  horizontalJustification?: Union<typeof TextHorizontalJustification>;
  verticalJustification?: Union<typeof TextVerticalJustification>;
  alignmentPoint?: Point3D;
  extrusion: Point3D;
  insert: Insert;

  override get subClassMarker(): string | undefined {
    return "AcDbText";
  }

  override get changeOwner(): boolean {
    return false;
  }

  constructor(options: AttribOptions, handle: Handle) {
    super("ATTRIB", handle, options);
    this.thickness = options.thickness;
    this.startPoint = options.startPoint;
    this.height = options.height;
    this.value = options.value;
    this.tag = options.tag;
    this.flags = options.flags ?? 0;
    this.rotation = options.rotation;
    this.relativeXScaleFactor = options.relativeXScaleFactor;
    this.obliqueAngle = options.obliqueAngle;
    this.styleName = options.styleName;
    this.generationFlags = options.generationFlags;
    this.horizontalJustification = options.horizontalJustification;
    this.verticalJustification = options.verticalJustification;
    this.alignmentPoint = options.alignmentPoint;
    this.extrusion = options.extrusion || extrusion();
    this.insert = options.insert;
    this.ownerObjectHandle = options.insert.handleSeed;
  }

  protected override tagifyChild(mg: TagsManager): void {
    mg.add(39, this.thickness);
    mg.point(this.startPoint);
    mg.add(40, this.height);
    mg.add(1, this.value);
    mg.add(50, this.rotation);
    mg.add(41, this.relativeXScaleFactor);
    mg.add(51, this.obliqueAngle);
    mg.add(7, this.styleName);
    mg.add(100, "AcDbAttribute");
    mg.add(280, 0);
    mg.add(2, this.tag);
    mg.add(70, this.flags);
    mg.add(71, this.generationFlags);
    mg.add(72, this.horizontalJustification);
    mg.add(74, this.verticalJustification);
    mg.point(this.alignmentPoint, 1);
    mg.add(280, 1);
  }
}
