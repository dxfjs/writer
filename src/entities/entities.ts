import { Blocks } from "../blocks";
import { Taggable } from "../types";
import { TagsManager } from "../utils";

export class Entities implements Taggable {
  readonly blocks: Blocks;

  get modelSpace() {
    return this.blocks.modelSpace;
  }

  get paperSpace() {
    return this.blocks.paperSpace;
  }

  constructor(blocks: Blocks) {
    this.blocks = blocks;
  }

  tagify(mg: TagsManager): void {
    mg.sectionStart("ENTITIES");
    this.paperSpace.entities.forEach((e) => e.tagify(mg));
    this.modelSpace.entities.forEach((e) => e.tagify(mg));
    mg.sectionEnd();
  }
}
