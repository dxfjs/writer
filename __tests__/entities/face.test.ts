import { Face, InvisibleEdge } from "@/entities";
import { Handle, TagsManager, point } from "@/utils";

describe("Face class", () => {
  it("should cover the face entity", () => {
    const mg = new TagsManager();
    const handle = new Handle();
    const face1 = new Face(
      {
        first: point(),
        second: point(10),
        third: point(10, 10),
        fourth: point(0, 10),
      },
      handle
    );
    face1.tagify(mg);
    const face2 = new Face(
      {
        first: point(0, 0, 10),
        second: point(10, 0, 10),
        third: point(10, 10, 10),
      },
      handle
    );
    face2.tagify(mg);

    const face3 = new Face(
      {
        first: point(0, 0, 5),
        second: point(10, 0, 5),
        third: point(10, 10, 5),
        fourth: point(0, 10, 5),
        flags: InvisibleEdge.Third,
      },
      handle
    );
    face3.tagify(mg);
    expect(mg.stringify()).toMatchSnapshot();
  });
});
