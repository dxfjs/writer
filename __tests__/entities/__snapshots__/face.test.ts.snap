// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Face class > should cover the face entity 1`] = `
"0
3DFACE
5
1
330
0
100
AcDbEntity
8
0
100
AcDbFace
10
0
20
0
30
0
11
10
21
0
31
0
12
10
22
10
32
0
13
0
23
10
33
0
70
0
0
3DFACE
5
2
330
0
100
AcDbEntity
8
0
100
AcDbFace
10
0
20
0
30
10
11
10
21
0
31
10
12
10
22
10
32
10
13
10
23
10
33
10
70
0
0
3DFACE
5
3
330
0
100
AcDbEntity
8
0
100
AcDbFace
10
0
20
0
30
5
11
10
21
0
31
5
12
10
22
10
32
5
13
0
23
10
33
5
70
4"
`;
